odoo/addons/base/models/ir_async.py

```diff
@@ -0,0 +1,265 @@
+        # somewhere. The async_traceback is a technical artifact so we can
+        # retrieve the complete traceback in case we re-queue an async job.
+        job['context']['async_traceback'] = job['traceback']
+        job['context']['async_job_id'] = job['id']
+
+        env = api.Environment(job_cr, job['user_id'], job['context'], job['super_user'])
+        ir_async = env['ir.async']
+
+        ir_async._pre_process(job)
+
+        try:
+            _logger.info('[%d] Calling "%s" on "%s"', job['id'], job['method'], job['res_model'])
+            with job_cr.savepoint():
+                records = env[job['res_model']].browse(job['res_ids'])
+                result = getattr(records, job['method'])(*job['args'], **job['kwargs'])
+                json.dumps(result)  # ensure result is serializable
+        except Exception as exc:
+            job['state'] = 'failed'
+            job['payload'] = ir_async._handle_exception(exc, job['id'])
+        else:
+            if result is None:
+                _logger.debug("[%d] Done", job['id'])
+                job['state'] = 'done'
+            else:
+                _logger.debug("[%d] Succeeded, result is %s", job['id'], result)
+                job['state'] = 'succeeded'
+                job['payload'] = ir_async._json_response(result)
+
+        ir_async._post_process(job)
+
+    def _pre_process(self, job):
+        pass
+
+    def _post_process(self, job):
+        pass
+
+    def _merge_traceback(self, tb):
+        """ Merge the given traceback with all previous levels of execution """
+        old_tb = self.env.context.get('async_traceback', '').splitlines(keepends=True)
+        if old_tb:
+            def fnc_process_job_not_reached(line):
+                return not re.search(f'in {self._process_job.__name__}\\W', line)
+            old_tb.append("Traceback of async job (most recent call last):\n")
+            old_tb.extend(itertools.dropwhile(fnc_process_job_not_reached, tb))
+            tb = old_tb
+        return "".join(tb)
+
+    def _handle_exception(self, exception, job_id):
+        header, *tb = traceback.format_exc().splitlines(keepends=True)
+        exc_tb = header + self._merge_traceback(tb)
+
+        if isinstance(exception, odoo.exceptions.UserError):
+            _logger.warning("[%d] %s", job_id, exception)
+        else:
+            _logger.error("[%d] Failed to process async job\n%s", job_id, exc_tb)
+
+        error = {
+            'code': 200,
+            'message': "Odoo Server Error",
+            'data': serialize_exception(exception),
+        }
+        error['data']['debug'] = exc_tb
+        return self._json_response(error=error)
+
+    def _json_response(self, result=None, error=None):
+        response = {}
+        if result is not None:
+            response['result'] = result
+        if error is not None:
+            response['error'] = error
+        return response
+
+    @api.autovacuum
+    def _vacuum_terminated_tasks(self):
```

Comment:
`terminated` is not a state of ir_async, and is strange terminology: it implies that the job was somehow killed during execution, not that it executed to completion (or failed).